pkg/controller/networkpolicy/networkpolicy_controller_query_test.go

```diff
@@ -0,0 +1,230 @@
+	querier := NewEndpointQuerier(c.NetworkPolicyController)
+	// start informers and run controller
+	c.informerFactory.Start(stopCh)
+	go c.Run(stopCh)
+	// wait until computation is done, we assume it is done when no signal has been received on heartbeat channel for 500ms
+	idleTimeout := 500 * time.Millisecond
+	timer := time.NewTimer(idleTimeout)
+	func() {
+		for {
+			timer.Reset(idleTimeout)
+			select {
+			case <-c.heartbeatCh:
+				continue
+			case <-timer.C:
+				close(stopCh)
+				return
+			}
+		}
+	}()
+	// block until computation complete
+	<-stopCh
+	return querier
+}
+
+// TestInvalidSelector tests the result of QueryNetworkPolicy when the selector (right now pod, namespace) does not
+// select any pods
+func TestInvalidSelector(t *testing.T) {
+	endpointQuerier := makeControllerAndEndpointQuerier()
+	// test appropriate response to QueryNetworkPolices
+	namespace, pod := "non-existing-namespace", "non-existing-pod"
+	response, err := endpointQuerier.QueryNetworkPolicies(namespace, pod)
+	if response != nil {
+		assert.Fail(t, "expected nil endpoints")
+	}
+	assert.Equal(t, nil, err, "expected not nil error")
+}
+
+// TestNoPolicy tests the result of QueryNetworkPolicy when no policies are relevant to the endpoint
+func TestNoPolicy(t *testing.T) {
+	endpointQuerier := makeControllerAndEndpointQuerier(&namespaces[0], &pods[0])
+	namespace1, pod1 := "testNamespace", "podA"
+	response1, err := endpointQuerier.QueryNetworkPolicies(namespace1, pod1)
+	require.Equal(t, nil, err)
+	// test applied policy response
+	assert.Equal(t, 0, len(response1.Endpoints[0].Policies))
+	// test egress + ingress policy response
+	assert.Equal(t, 0, len(response1.Endpoints[0].Rules))
+}
+
+// TestSingleAppliedPolicy tests the result of QueryNetworkPolicy when the selector (right now pod, namespace) selects a
+// pod which has a single network policy object applied to it
+func TestSingleAppliedIngressEgressPolicy(t *testing.T) {
+	endpointQuerier := makeControllerAndEndpointQuerier(&namespaces[0], &pods[0], &policies[0])
+	namespace1, pod1 := "testNamespace", "podA"
+	response1, err := endpointQuerier.QueryNetworkPolicies(namespace1, pod1)
+	require.Equal(t, nil, err)
+	// test applied policy response
+	assert.Equal(t, "test-ingress-egress", response1.Endpoints[0].Policies[0].PolicyRef.Name)
+	// test egress policy response
+	assert.Equal(t, v1beta1.DirectionOut, response1.Endpoints[0].Rules[0].Direction)
+	assert.Equal(t, "test-ingress-egress", response1.Endpoints[0].Rules[0].PolicyRef.Name)
+	// test ingress policy response
+	assert.Equal(t, v1beta1.DirectionIn, response1.Endpoints[0].Rules[1].Direction)
+	assert.Equal(t, "test-ingress-egress", response1.Endpoints[0].Rules[1].Name)
+}
+
+// TestMultiplePolicy tests the result of QueryNetworkPolicy when the selector (right now pod, namespace) selects
+// a pod which has multiple networkpolicies which define policies on it.
+func TestMultiplePolicy(t *testing.T) {
```

Comment:
Feel the 4 tests could be in single table-driven test to reduce redundancy, but of course it could be improved later.

Reply:
Thanks for the review. I modified the tests as you suggested.